File: Software/Modules/Arduino/main/src/domain/dimmer.cpp
#include "dimmer.h"

namespace gw {

namespace {

bool hex_nibble(char c, uint8_t& out)
{
    if (c >= '0' && c <= '9') {
        out = static_cast<uint8_t>(c - '0');
        return true;
    }
    if (c >= 'a' && c <= 'f') {
        out = static_cast<uint8_t>(c - 'a' + 10);
        return true;
    }
    if (c >= 'A' && c <= 'F') {
        out = static_cast<uint8_t>(c - 'A' + 10);
        return true;
    }
    return false;
}

bool hex_byte(const char* s, uint8_t& out)
{
    uint8_t hi = 0;
    uint8_t lo = 0;
    if (!hex_nibble(s[0], hi) || !hex_nibble(s[1], lo)) {
        return false;
    }
    out = static_cast<uint8_t>((hi << 4) | lo);
    return true;
}

uint8_t length_of(const char* s)
{
    uint8_t n = 0;
    while (s != nullptr && s[n] != '\0' && n < 255) {
        ++n;
    }
    return n;
}

} // namespace

Dimmer::Dimmer(IPwm& pwm, IClock& clock, const Pin (&pins)[kMaxDimmerChannels], uint8_t channels,
               const DimmerTuning& tuning)
    : pwm_(pwm), clock_(clock), channels_(channels > kMaxDimmerChannels ? kMaxDimmerChannels
                                                                       : channels),
      tuning_(tuning)
{
    for (uint8_t i = 0; i < kMaxDimmerChannels; ++i) {
        pins_[i] = pins[i];
    }
}

void Dimmer::begin()
{
    for (uint8_t i = 0; i < channels_; ++i) {
        pwm_.write_duty(pins_[i], 0);
    }
    on_ = false;
}

uint8_t Dimmer::channel_value(uint8_t channel) const
{
    return channel < kMaxDimmerChannels ? values_[channel] : 0;
}

void Dimmer::write_outputs()
{
    for (uint8_t i = 0; i < channels_; ++i) {
        const uint16_t scaled =
            static_cast<uint16_t>(static_cast<uint32_t>(level_) * values_[i] / 100u);
        pwm_.write_duty(pins_[i], static_cast<uint8_t>(scaled > 255 ? 255 : scaled));
    }
}

bool Dimmer::step_towards(uint8_t& current, uint8_t target, uint8_t step)
{
    if (current == target) {
        return false;
    }
    const uint8_t effective = step == 0 ? 1 : step;
    if (current < target) {
        const uint8_t room = static_cast<uint8_t>(target - current);
        current = static_cast<uint8_t>(current + (effective < room ? effective : room));
    } else {
        const uint8_t room = static_cast<uint8_t>(current - target);
        current = static_cast<uint8_t>(current - (effective < room ? effective : room));
    }
    return true;
}

void Dimmer::set_target_level(uint8_t percent)
{
    target_level_ = percent > 100 ? 100 : percent;
}

void Dimmer::bump_level(uint8_t step)
{
    const uint16_t next = static_cast<uint16_t>(target_level_) + step;
    target_level_ = next > 100 ? step : static_cast<uint8_t>(next);
}

void Dimmer::set_on(bool on)
{
    if (on == on_) {
        return;
    }
    on_ = on;

    // The requested brightness must survive an off/on cycle, so it is restored
    // after the ramp rather than being used as the ramp target directly.
    const uint8_t requested = target_level_;

    if (on) {
        level_ = 0;
        target_level_ = requested;
        // Colours are unknown at switch-on; adopt the targets so the ramp has
        // something to scale.
        for (uint8_t i = 0; i < channels_; ++i) {
            if (values_[i] == 0) {
                values_[i] = target_values_[i];
            }
        }
        update();
    } else {
        target_level_ = 0;
        update();
        level_ = requested; // remember brightness for the next switch-on
    }

    target_level_ = requested;
}

bool Dimmer::set_colors_from_hex(const char* hex)
{
    if (hex == nullptr) {
        return false;
    }

    const char* p = hex;
    uint8_t len = length_of(p);
    if (len > 0 && p[0] == '#') {
        ++p;
        --len;
    }
    if (len != 6 && len != 8) {
        return false;
    }

    const uint8_t available = static_cast<uint8_t>(len / 2);
    uint8_t parsed[kMaxDimmerChannels] = {0, 0, 0, 0};
    for (uint8_t i = 0; i < available && i < kMaxDimmerChannels; ++i) {
        if (!hex_byte(p + 2 * i, parsed[i])) {
            return false; // reject atomically; do not half-apply a bad payload
        }
    }

    for (uint8_t i = 0; i < channels_ && i < available; ++i) {
        target_values_[i] = parsed[i];
    }
    return true;
}

void Dimmer::update()
{
    if (!on_) {
        return;
    }

    uint32_t next_step_at = clock_.now_ms();

    for (;;) {
        const uint32_t now = clock_.now_ms();
        if (now < next_step_at) {
            // millis() rollover: resynchronise rather than stalling for 49 days.
            next_step_at = now;
            continue;
        }
        if (now - next_step_at < tuning_.interval_ms) {
            continue;
        }

        bool moved = step_towards(level_, target_level_, tuning_.step);
        for (uint8_t i = 0; i < channels_; ++i) {
            moved = step_towards(values_[i], target_values_[i], tuning_.step) || moved;
        }

        write_outputs();
        next_step_at = clock_.now_ms();

        if (!moved) {
            return;
        }
    }
}

} // namespace gw

File: Software/Modules/Arduino/rust_src/crates/gowired-core/src/domain/dimmer.rs
//! PWM dimmer with white / RGB / RGBW colour models.
//!
//! A port of GoWired-lib's `Dimmer` class. Two defects in the original are fixed
//! here and covered by tests:
//!
//! - Hex colour parsing only accepted upper-case digits. `_StringHexToByte()`
//!   did `c -= 7` for anything above `'9'`, which is right for `'A'..'F'` and off
//!   by 32 for `'a'..'f'`, so `"ff0000"` decoded to garbage. Controllers
//!   generally send lower case -- and the original's own init confirmation
//!   advertised `"ffffff"`. Parsing is case-insensitive here and accepts an
//!   optional `'#'`.
//!
//! - The ramp loops compared `current != target` while stepping by
//!   `DimmerTuning::step`, so any step size that did not exactly divide the
//!   distance overshot and spun forever until the watchdog fired. The final step
//!   is clamped to the remaining distance.

use crate::domain::config::DimmerTuning;
use crate::hal::{Clock, Pin, Platform, Pwm, NO_PIN};

/// Maximum channels a dimmer can drive (R, G, B, W).
pub const MAX_CHANNELS: usize = 4;

fn hex_nibble(c: u8) -> Option<u8> {
    match c {
        b'0'..=b'9' => Some(c - b'0'),
        b'a'..=b'f' => Some(c - b'a' + 10),
        b'A'..=b'F' => Some(c - b'A' + 10),
        _ => None,
    }
}

fn hex_byte(s: &[u8]) -> Option<u8> {
    let hi = hex_nibble(s[0])?;
    let lo = hex_nibble(s[1])?;
    Some((hi << 4) | lo)
}

/// Brightness and colour ramp over up to four PWM channels.
pub struct Dimmer<'a, P: Platform> {
    pwm: &'a P::Pwm,
    clock: &'a P::Clock,
    pins: [Pin; MAX_CHANNELS],
    channels: usize,
    tuning: DimmerTuning,

    on: bool,
    level: u8,
    target_level: u8,
    values: [u8; MAX_CHANNELS],
    target_values: [u8; MAX_CHANNELS],
}

impl<'a, P: Platform> Dimmer<'a, P> {
    /// Wires a dimmer to `channels` of the given pins.
    pub const fn new(
        pwm: &'a P::Pwm,
        clock: &'a P::Clock,
        pins: [Pin; MAX_CHANNELS],
        channels: u8,
        tuning: DimmerTuning,
    ) -> Self {
        let channels = if channels as usize > MAX_CHANNELS {
            MAX_CHANNELS
        } else {
            channels as usize
        };
        Self {
            pwm,
            clock,
            pins,
            channels,
            tuning,
            on: false,
            level: 20,
            target_level: 20,
            values: [0; MAX_CHANNELS],
            target_values: [255; MAX_CHANNELS],
        }
    }

    /// An unwired dimmer, for a device that has none.
    pub const fn disabled(pwm: &'a P::Pwm, clock: &'a P::Clock, tuning: DimmerTuning) -> Self {
        Self::new(pwm, clock, [NO_PIN; MAX_CHANNELS], 0, tuning)
    }

    /// Drives every channel to zero.
    pub fn begin(&mut self) {
        for i in 0..self.channels {
            self.pwm.write_duty(self.pins[i], 0);
        }
        self.on = false;
    }

    /// Whether the strip is switched on.
    pub const fn is_on(&self) -> bool {
        self.on
    }

    /// Brightness the dimmer is ramping towards, 0..100.
    pub const fn target_level(&self) -> u8 {
        self.target_level
    }

    /// Brightness currently applied to the outputs, 0..100.
    pub const fn level(&self) -> u8 {
        self.level
    }

    /// Current value of one colour channel, 0..255.
    pub fn channel_value(&self, channel: usize) -> u8 {
        if channel < MAX_CHANNELS {
            self.values[channel]
        } else {
            0
        }
    }

    fn write_outputs(&self) {
        for i in 0..self.channels {
            let scaled = u32::from(self.level) * u32::from(self.values[i]) / 100;
            self.pwm
                .write_duty(self.pins[i], if scaled > 255 { 255 } else { scaled as u8 });
        }
    }

    /// Steps `current` towards `target` by at most `step`; true if it moved.
    fn step_towards(current: &mut u8, target: u8, step: u8) -> bool {
        if *current == target {
            return false;
        }
        let effective = if step == 0 { 1 } else { step };
        if *current < target {
            let room = target - *current;
            *current += effective.min(room);
        } else {
            let room = *current - target;
            *current -= effective.min(room);
        }
        true
    }

    /// `V_PERCENTAGE` from the controller. Values above 100 are clamped.
    pub fn set_target_level(&mut self, percent: u8) {
        self.target_level = percent.min(100);
    }

    /// Wall-switch brightness step. Wraps back to `step` once past 100.
    pub fn bump_level(&mut self, step: u8) {
        let next = u16::from(self.target_level) + u16::from(step);
        self.target_level = if next > 100 { step } else { next as u8 };
    }

    /// Turns the dimmer on or off, ramping the brightness.
    pub fn set_on(&mut self, on: bool) {
        if on == self.on {
            return;
        }
        self.on = on;

        // The requested brightness must survive an off/on cycle, so it is
        // restored after the ramp rather than used as the ramp target directly.
        let requested = self.target_level;

        if on {
            self.level = 0;
            self.target_level = requested;
            // Colours are unknown at switch-on; adopt the targets so the ramp
            // has something to scale.
            for i in 0..self.channels {
                if self.values[i] == 0 {
                    self.values[i] = self.target_values[i];
                }
            }
            self.update();
        } else {
            self.target_level = 0;
            self.update();
            self.level = requested; // remember brightness for the next switch-on
        }

        self.target_level = requested;
    }

    /// Parses `"RRGGBB"` or `"RRGGBBWW"`, optionally `'#'`-prefixed,
    /// case-insensitive.
    ///
    /// Returns `false` if the payload was not valid hex, leaving the colours
    /// untouched.
    pub fn set_colors_from_hex(&mut self, hex: &str) -> bool {
        let mut bytes = hex.as_bytes();
        if bytes.first() == Some(&b'#') {
            bytes = &bytes[1..];
        }
        if bytes.len() != 6 && bytes.len() != 8 {
            return false;
        }

        let available = bytes.len() / 2;
        let mut parsed = [0u8; MAX_CHANNELS];
        for i in 0..available.min(MAX_CHANNELS) {
            match hex_byte(&bytes[2 * i..]) {
                Some(b) => parsed[i] = b,
                // Reject atomically; do not half-apply a bad payload.
                None => return false,
            }
        }

        let n = self.channels.min(available);
        self.target_values[..n].copy_from_slice(&parsed[..n]);
        true
    }

    /// Advances the brightness/colour ramp to the target.
    ///
    /// Blocking, as in the original: returns once the outputs have reached the
    /// requested values.
    pub fn update(&mut self) {
        if !self.on {
            return;
        }

        let mut next_step_at = self.clock.now_ms();

        loop {
            let now = self.clock.now_ms();
            if now < next_step_at {
                // millis() rollover: resynchronise rather than stalling for 49
                // days.
                next_step_at = now;
                continue;
            }
            if now - next_step_at < u32::from(self.tuning.interval_ms) {
                continue;
            }

            let mut moved = Self::step_towards(&mut self.level, self.target_level, self.tuning.step);
            for i in 0..self.channels {
                let target = self.target_values[i];
                moved = Self::step_towards(&mut self.values[i], target, self.tuning.step) || moved;
            }

            self.write_outputs();
            next_step_at = self.clock.now_ms();

            if !moved {
                return;
            }
        }
    }
}
